lib/chef-dk/policyfile/git_lock_fetcher.rb

```diff
@@ -0,0 +1,248 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "chef-dk/policyfile_lock"
+require "chef-dk/exceptions"
+require "chef-dk/helpers"
+require "mixlib/shellout"
+require "tmpdir"
+
+module ChefDK
+  module Policyfile
+
+    # A Policyfile lock fetcher that can read a lock file from a git repository.
+    #
+    # @author Ryan Hass
+    # @author Daniel DeLeo
+    #
+    # @since 3.1.0
+    #
+    class GitLockFetcher
+      attr_accessor :name
+      attr_accessor :source_options
+      attr_accessor :storage_config
+
+      attr_reader :uri
+      attr_reader :branch
+      attr_reader :tag
+      attr_reader :ref
+      attr_reader :revision
+      attr_reader :path
+
+      # Initialize a GitLockFetcher
+      #
+      # @param name [String] The name of the policyfile
+      # @param source_options [Hash] A hash with a :path key pointing at the location
+      #                              of the lock
+      def initialize(name, source_options, storage_config)
+        @name           = name
+        @storage_config = storage_config
+        @source_options = source_options
+
+        @uri      = source_options[:git]
+        @branch   = source_options[:branch]
+        @tag      = source_options[:tag]
+        @ref      = source_options[:ref]
+        @revision = source_options[:revision]
+        @path     = source_options[:path] || source_options[:rel]
+
+        # The revision to parse
+        @rev_parse = source_options[:ref] || source_options[:branch] || source_options[:tag] || "master"
+      end
+
+      # @return [True] if there were no errors with the provided source_options
+      # @return [False] if there were errors with the provided source_options
+      def valid?
+        errors.empty?
+      end
+
+      # Check the options provided when craeting this class for errors
+      #
+      # @return [Array<String>] A list of errors found
+      def errors
+        error_messages = []
+        [:git].each do |key|
+          error_messages << "include_policy for #{name} is missing key #{key}" unless source_options[key]
+        end
+
+        error_messages
+      end
+
+      # @return [Hash] The source_options that describe how to fetch this exact lock again
+      def source_options_for_lock
+        source_options.merge({
+                               revision:  revision,
+                             })
+      end
+
+      # Applies source options from a lock file. This is used to make sure that the same
+      # policyfile lock is loaded that was locked
+      #
+      # @param options_from_lock [Hash] The source options loaded from a policyfile lock
+      def apply_locked_source_options(options_from_lock)
+        # There are no options the lock could provide
```

Comment:
where does revision get loaded? I kind of expected this to be filled out if `source_options_for_lock` was

Reply:
resolved now